### PR TITLE
Literal processing and type adaptation

### PR DESCRIPTION
**Pull request by wiskojo**
### Motivation
The Teradata dialect currently does not support literal processing and type adaptation. This PR provides this functionality and would allow users to execute arithmetic operations of Python literals against Teradata data types. Additionally, the general resulting type of such operations can be inferred at the application level through the `BinaryExpression`'s `type` field.

### Description of Changes:
- Implemented `literal_processor` for each SQLAlchemy data type corresponding to a Teradata literal.
- Defined mapping from Python literal to SQLAlchemy data type in `coerce_compared_value`.
- Separated out different strategies for doing expression adaptation and implemented a new `_IntervalRuleStrategy` for inferring Interval expression type.
- Added integration tests to check for the correctness of literal type inference.
- Added unit tests to check that the supported Python literals render to the correct syntax corresponding to a Teradata literal.
- Slightly modified `CreateView` to explicitly compile with `literal_binds=True` and implemented a basic version of `CreateTableAs`.

## Related Issues
NA

## Reviewers
- @sandan 

# CHECKLIST:
Make sure all items are marked when you submit the pull-request.

- [x] Relevant documentation for functions, tests, classes, the wiki, etc. have been made
- [x] Necessary unit tests in tests/ pass with no errors
- [x] Necessary integration tests in tests/ pass with no errors
- [ ] Update the CHANGELOG.md with a summary of your changes if requested

### CODE REVIEW

File: sqlalchemy_teradata/types.py [hunk 1]
```diff
@@ -8,39 +8,86 @@
...
+        elif type_ == decimal.Decimal:
+            return DECIMAL()
+        elif type_ == datetime.date:
+            return DATE()
+        elif type_ == datetime.datetime:
+            return TIMESTAMP()
+        elif type_ == datetime.time:
+            return TIME()
+        elif type_ == td_dtypes.Interval:
+            return globals().get(
```
Comment (sandan) [inline sqlalchemy_teradata/types.py]: Consider limiting the scope of the call when getting sqlalchemy Interval type.

File: sqlalchemy_teradata/types.py [hunk 2]
```diff
@@ -289,6 +333,12 @@ def __init__(self, **kwargs):
         """ Construct a FLOAT object """
         super(FLOAT, self).__init__(**kwargs)
 
+    def literal_processor(self, dialect):
+
+        def process(value):
```
Comment (sandan) [inline sqlalchemy_teradata/types.py:338]: Quote the manual to say why `.14e` is chosen.

Comment (sandan) [reply sqlalchemy_teradata/types.py:338]: Also choose the min (precision of python float, precision database float)

Comment (sandan) [reply sqlalchemy_teradata/types.py:338]: Use a CAST to cast python float literals to FLOAT.

File: sqlalchemy_teradata/types.py [hunk 3]
```diff
@@ -1010,196 +1100,333 @@ def process(self, type_, op=None, other=None, **kw):
...
+        #
+        # E.g. INTERVAL YEAR TO MONTH and INTERVAL MINUTE TO SECOND -->
+        #      INTERVAL YEAR TO SECOND (invalid) -->
+        #      NullType
+        elif isinstance(type_, _TDInterval) and isinstance(other, _TDInterval):
+            tokens = self._tokenize_name(type_.__class__.__name__) + \
+                     self._tokenize_name(other.__class__.__name__)
+            tokens.sort(key=lambda tok: self.ordering.index(tok))
+
+            return globals().get(
```
Comment (sandan) [inline sqlalchemy_teradata/types.py]: See comment above about globals()

File: sqlalchemy_teradata/types.py [hunk 4]
```diff
@@ -1010,196 +1100,333 @@ def process(self, type_, op=None, other=None, **kw):
...
+                    flat_dict[k] = v
+            else:
+                flat_dict[ks] = v
         return flat_dict
 
-    def visit_INTEGER(self, type_, **kw):
-        return self._flatten_tuple_dict({
-            (operators.add, operators.sub, operators.mul,
-             operators.truediv, operators.mod): {
+    def visit_INTEGER(self, type_, other, **kw):
```
Comment (sandan) [inline sqlalchemy_teradata/types.py:1249]: Be sure to mention the default type being the left operand (i.e. the name of the visit type method)

**Review by sandan — APPROVED** [review]
Please see comments.

File: sqlalchemy_teradata/types.py (hunk 1 above)
Comment (wiskojo) [reply sqlalchemy_teradata/types.py]: For now resolved by changing the call to `globals()` to `sys.modules` instead.

File: sqlalchemy_teradata/types.py (hunk 4 above)
Comment (wiskojo) [reply sqlalchemy_teradata/types.py:1249]: Will add this in the documentation/wiki page.

File: sqlalchemy_teradata/types.py (hunk 2 above)
Comment (wiskojo) [reply sqlalchemy_teradata/types.py:338]: Now using `CAST(%s as FLOAT)` instead of `%.14e`.

**Pull request merged**